sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -10,3 +10,175 @@
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, provisioning_service_api_version, &mqtt_user_name));
+
+  if (az_span_length(*user_agent) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_user_name, '&', &mqtt_user_name));
+    AZ_RETURN_IF_FAILED(az_span_append(mqtt_user_name, *user_agent, &mqtt_user_name));
+  }
+
+  *out_mqtt_user_name = mqtt_user_name;
+  return AZ_OK;
+}
+
+// <registration_id>
+AZ_NODISCARD az_result az_iot_provisioning_client_id_get(
+    az_iot_provisioning_client const* client,
+    az_span mqtt_client_id,
+    az_span* out_mqtt_client_id)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_client_id, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_client_id);
+
+  AZ_RETURN_IF_FAILED(
+      az_span_copy(mqtt_client_id, client->_internal.registration_id, &mqtt_client_id));
+
+  *out_mqtt_client_id = mqtt_client_id;
+  return AZ_OK;
+}
+
+// $dps/registrations/res/#
+AZ_NODISCARD az_result az_iot_provisioning_client_register_subscribe_topic_filter_get(
+    az_iot_provisioning_client const* client,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter)
+{
+  UNUSED(client);
+
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic_filter, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic_filter);
+
+  AZ_RETURN_IF_FAILED(az_span_copy(mqtt_topic_filter, str_dps, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, str_registrations, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, str_res, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_topic_filter, '#', &mqtt_topic_filter));
+
+  *out_mqtt_topic_filter = mqtt_topic_filter;
+  return AZ_OK;
+}
+
+// topic: $dps/registrations/res/202/?$rid=1&retry-after=3
+// payload:
+// {"operationId":"4.d0a671905ea5b2c8.42d78160-4c78-479e-8be7-61d5e55dac0d","status":"assigning"}
+AZ_NODISCARD az_result az_iot_provisioning_client_received_topic_payload_parse(
+    az_iot_provisioning_client const* client,
+    az_span received_topic,
+    az_span received_payload,
+    az_iot_provisioning_client_register_response* out_response)
+{
+  UNUSED(client);
+  UNUSED(received_topic);
+  UNUSED(received_payload);
+  UNUSED(out_response);
+
+  return AZ_ERROR_NOT_IMPLEMENTED;
+}
+
+// $dps/registrations/PUT/iotdps-register/?$rid={%s}
+AZ_NODISCARD az_result az_iot_provisioning_client_register_publish_topic_get(
+    az_iot_provisioning_client const* client,
+    az_span mqtt_topic,
+    az_span* out_mqtt_topic)
+{
+  UNUSED(client);
+
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic);
+
+  AZ_RETURN_IF_FAILED(az_span_copy(mqtt_topic, str_dps, &mqtt_topic));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic, str_registrations, &mqtt_topic));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic, str_put_iotdps_register, &mqtt_topic));
```

Comment:
It looks like to be a constant, can we have one rather then computing it?

Reply:
We're investigating that offline. Not part of this PR.